### PR TITLE
CD(): playwright codegen

### PR DESCRIPTION
**Pull request by ShaMan123**
## Motivation

We want tests to be easy to write.
I saw you commented on this https://github.com/microsoft/playwright/issues/21504 that requests support for mouse events etc.
I looked into it and decided to impl a simple solution for fabric

## Description

setupCodegen is fabric's version for playwright's codegen.
It adds event listeners and generates playwright code when the test is done and attachs it to the test results.

### Usage

- start the template ` npm run test:e2e -- e2e/tests/template/index.spec.ts`
- in the chrome devtools exec `startRecording()`
- play with the canvas
- add a step by calling `step('step name')`
- take a screenshot by calling `captureScreenshot()`
- hit the play button of the playwright devtools
- the generated code will be logged to the console and attached to the test results

## In Action

image

### CODE REVIEW

**Review by ShaMan123 — COMMENTED** [review]
This is a game changer

File: package.json [hunk 1]
```diff
@@ -144,6 +144,12 @@
       ],
       "src/*": [
         "src/*"
+      ],
+      "test/playwright": [
+        "e2e/setup/index.ts"
+      ],
+      "test/jest": [
+        "jest.setup.ts"
```
Comment (ShaMan123) [inline package.json:152]: now consumers can use it

Comment (ShaMan123) [reply package.json:152]: but I don't knowe how to make typescipt auto complete `fabric/tes` => `fabric/test/..`

**Review by ShaMan123 — COMMENTED** [review]
I love this.
Testing will become a lot easier with this.

**Comment by ShaMan123** [comment]
Now that I understand better playwright I think we should configure fixtures if we can, instead on before/after hooks

**Comment by asturur** [comment]
Sorry  i was sure i posted a comment here weeks ago.
Can't this become a plugin and leave as its own library? isn't this reusable outside fabricJS?

**Comment by ShaMan123** [comment]
I guess,
So then I install it on the repo?

**Comment by ShaMan123** [comment]
Do you want it on the org or shall I put it on my account?

**Comment by asturur** [comment]
In my opinion this is very useful for brush tests or generic test for which we can't use center of objects or text positions.
In my opinion you should make a plugin, your repo or the org repo is your preference.
There is chance this is very useful to more people.

**Comment by ShaMan123** [comment]
closed in favor of https://github.com/ShaMan123/playwright-codegen

**Pull request closed without merging**